### PR TITLE
Cnam 418: Interaction Transformer

### PR DESCRIPTION
**Pull request by ysebiat**
This the first Transformer of Interactions. An Interaction Transformer combines multiple Exposures to form a higher level of Exposure where the number of values(a value can a Molecule, Pharmaceutical or Therapeutic depending on the configuration level in the study) is superior to 2. 

An Interaction uses the weight to express the Level. The Level is the number of values involved in forming the Interaction. Thus, an Exposure is an interaction of Level 1 (1 value involved)

In the current implementation, there is only one configuration control which is the "level". "Level" is the highest Interaction Level that the user wishes to produce. Thus, if the data does not contain the "level", it will produce the highest possible Interaction Level the data contains.

The current implementation is sensitive to data skew over a patient. If the data contains a patient with a very high level of Exposures, going beyond Level 5 can be tricky.

This PR, includes our first attempt at Functional programming. We use the Cats library to do so.

### CODE REVIEW

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/interaction/NLevelInteractionTransformer.scala [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    ens.map(dataset => dataset.flatMap(e => e.toLowerLevelInvolvedExposureN))
+  }
+
+  def reduceHigherExposuresNFromLowerExposures(higher: List[Dataset[ExposureN]], lower: List[Dataset[ExposureN]])
+  : List[Dataset[LeftRemainingPeriod[ExposureN]]] = {
+    val sqlCtx = higher.head.sqlContext
+    import sqlCtx.implicits._
+    lower.zip(higher).map(
+      e => {
```
Comment (dsun0720) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/interaction/NLevelInteractionTransformer.scala]: use (lower, higher)as pattern

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/interaction/NLevelInteractionTransformer.scala [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        val down1 = e._1
+        val high = e._2
+        high.joinWith(
+          down1,
+          high("patientID") === down1("patientID") && high("values") === down1("values"),
+          "left"
+        ).groupByKey(e => e._1)
+          .flatMapGroups(
+            (e, i) =>
```
Comment (dsun0720) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/interaction/NLevelInteractionTransformer.scala]: e, and i is not clear, try to use meaningful variable like: (higer, iter)

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/interaction/NLevelInteractionTransformer.scala [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+              RemainingPeriod.delimitPeriods(
+                RightRemainingPeriod(e),
+                i.map(_._2).toList.sortBy(_.period.start).map(e => LeftRemainingPeriod[ExposureN](e)),
```
Comment (dsun0720) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/interaction/NLevelInteractionTransformer.scala]: in fact, i is an iterator of (higher, lower), try to use this pattern in the map

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/interaction/NLevelInteractionTransformer.scala [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+  def joinTwoExposureNDataSet(right: Dataset[ExposureN], left: Dataset[ExposureN]): Dataset[ExposureN] = {
+    val sqlCtx = right.sqlContext
+    import sqlCtx.implicits._
+    right
+      .joinWith(
+        left,
+        left(Event.Columns.PatientID) === right(Event.Columns.PatientID) && !left("values").geq(right("values"))
+      )
+      .flatMap(e => e._1.intersect(e._2))
```
[the remark this answers is not in the document]
Comment (ysebiat) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/interaction/NLevelInteractionTransformer.scala]: You cannot unpack a tuple in that way in Scala.

**Comment by ysebiat** [comment]
Here are my answers:

**Q**: Is there any None Value in exposure.end ? 
**A**: an Exposure with no End shouldn't exist and does not exist. I think the better solution is to enforce this by inheriting a trait that make sure the end is note None.

**Q**: I'm not sure spark.cache is necessary. I think there is only one action(write) called for this transformer, isn' t it ?
**A**: could you give me an example of this please?

**Comment by dsun0720** [comment]
I mean that cache is called several times in elevateToExposureN, but there is no action operation called in the whole transform. I don't understand why we need these caches.

**Comment by ysebiat** [comment]
Good point. Can you put the comment on the line of the cache so that i can give the reason?

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/interaction/NLevelInteractionTransformer.scala [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+      sqlCtx
+        .createDataset(exposures.map(e => ExposureN(e.patientID, Set(e.value), Period(e.start, e.end.get))).rdd)
+        .repartition(functions.col("patientID"), functions.col("values")).cache()
+    )
+
+    List
+      .fill(n - 1)(
+        sqlCtx
+          .createDataset(exposures.map(e => ExposureN(e.patientID, Set(e.value), Period(e.start, e.end.get))).rdd)
+          .repartition(functions.col("patientID"), functions.col("values")).cache()
```
Comment (dsun0720) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/transformers/interaction/NLevelInteractionTransformer.scala]: Does this cache have the effect to improve the performance during the join operation ?

**Pull request merged**